grouper/repositories/user.py

```diff
@@ -1,29 +1,97 @@
+from collections import defaultdict
 from typing import TYPE_CHECKING
 
-from grouper.entities.user import UserNotFoundException
-from grouper.models.user import User
+from grouper.entities.user import PublicKey, User, UserMetadata, UserNotFoundException
+from grouper.models.public_key import PublicKey as SQLPublicKey
+from grouper.models.user import User as SQLUser
+from grouper.models.user_metadata import UserMetadata as SQLUserMetadata
+from grouper.repositories.interfaces import UserRepository
 
 if TYPE_CHECKING:
+    from grouper.graph import GroupGraph
     from grouper.models.base.session import Session
+    from typing import Dict, List
 
 
-class UserRepository(object):
-    """Storage layer for users."""
+class GraphUserRepository(UserRepository):
+    """Graph-aware storage layer for users."""
+
+    def __init__(self, graph, repository):
+        # type: (GroupGraph, UserRepository) -> None
+        self.graph = graph
+        self.repository = repository
+
+    def all_enabled_users(self):
+        # type: () -> Dict[str, User]
+        return self.graph.all_user_metadata()
+
+    def disable_user(self, user):
+        # type: (str) -> None
+        self.repository.disable_user(user)
+
+    def user_is_enabled(self, name):
+        # type: (str) -> bool
+        """Return whether a user is enabled.
+
+        TODO(rra): This checks the underlying data store, not the graph, even though this
+        information is in the graph, because the convert_user_to_service_account usecase disables a
+        user and then immediately checks whether the user is disabled, and we don't want to force a
+        graph refresh in the middle of that usecase.  This indicates a deeper underlying problem
+        where some usecases need to use SQL repositories rather than the graph, which we're
+        deferring for future work when we significantly reorganize how Grouper uses the graph.
+        """
+        return self.repository.user_is_enabled(name)
+
+
+class SQLUserRepository(UserRepository):
+    """SQL storage layer for users."""
 
     def __init__(self, session):
         # type: (Session) -> None
         self.session = session
 
+    def all_enabled_users(self):
```

Comment:
Can we call this `all_users` or `get_all_users` instead?

Unnecessarily long explanation:
I've harped on this a few times but I want to be consistent with language around disabled things. I think the rest of the code under the new architecture has been settling on "disabled things are never included unless specified in function name or by param," making the `enabled` unnecessary. Moreover, I personally would definitely waste time if I were coming to this codebase, needed to (for example) get all enabled users and all enabled perms, and saw `all_enabled_users` vs. `list_permissions` since I'd want to go verify that I wasn't getting disabled perms in the results.

Reply:
Yeah, you're right. I confused myself again because the graph retains information for disabled users and I thought we'd later need a real `all_users`, but if we ever care about such a thing, we can provide a flag to this function saying to include disabled users. As is, right now, it just adds more confusion, and we should stick with an invariant where all the new APIs only return enabled things.